File: backend/app/schemas/sustainability.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime, date
from typing import Optional, List, Dict, Any, Generic, TypeVar
from enum import Enum
from uuid import UUID

T = TypeVar('T')

class ReportPeriod(str, Enum):
    DAILY = "DAILY"
    WEEKLY = "WEEKLY"
    MONTHLY = "MONTHLY"
    YEARLY = "YEARLY"
    CUSTOM = "CUSTOM"

class ReadinessLevel(str, Enum):
    NOT_READY = "NOT_READY"
    PARTIALLY_READY = "PARTIALLY_READY"
    READY = "READY"
    HIGHLY_READY = "HIGHLY_READY"

class RecommendationLevel(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"

# Standard JSON response envelope
class APIResponse(BaseModel, Generic[T]):
    success: bool = True
    message: str
    data: Optional[T] = None

# Requests
class CarbonCalculationRequest(BaseModel):
    vehicle_id: str
    start_date: date
    end_date: date
    grid_region: Optional[str] = "india"

class DieselVsEVRequest(BaseModel):
    distance_km: float
    payload_kg: float
    diesel_efficiency: float  # fuel consumption in L/100km
    ev_efficiency: float  # energy consumption in kWh/km
    vehicle_type: Optional[str] = "medium_truck"

class ReadinessAssessmentRequest(BaseModel):
    route_distance: float
    payload: float  # in kg (scorer works in tons, so we'll divide by 1000)
    charging_availability: bool
    dwell_time: float  # in hours
    vehicle_type: Optional[str] = "medium_truck"
    vehicle_id: Optional[str] = None

class ProcurementRecommendationRequest(BaseModel):
    fleet_size: int
    daily_distance: float
    charging_available: bool
    vehicle_type: Optional[str] = "medium_truck"

# Responses
class CarbonReportResponse(BaseModel):
    report_id: UUID
    vehicle_id: str
    distance_travelled: float
    energy_consumed_kwh: float
    diesel_emission: float
    ev_emission: float
    scope1_emission: float
    scope3_emission: float
    carbon_saved: float
    grid_region: str
    generated_at: datetime

    model_config = ConfigDict(from_attributes=True)

class DieselVsEVResponse(BaseModel):
    diesel_emission: float
    ev_emission: float
    carbon_saved: float
    reduction_percentage: float
    equivalent_trees: float
    equivalent_gallons_gasoline: float

class ReadinessAssessmentResponse(BaseModel):
    assessment_id: UUID
    readiness_score: float
    readiness_level: str
    recommendation: str
    factor_scores: Dict[str, float]
    financial: Dict[str, Any]
    improvements_needed: List[str]
    generated_at: datetime

    model_config = ConfigDict(from_attributes=True)

class ProcurementRecommendationResponse(BaseModel):
    recommended_vehicle_type: str
    recommended_quantity: int
    estimated_carbon_saving: float
    recommendation_level: RecommendationLevel
    reasoning: str

class SustainabilitySummaryResponse(BaseModel):
    total_carbon_saved_kg: float
    scope1_emission_kg: float
    scope3_emission_kg: float
    vehicles_assessed: int
    average_readiness_score: float
    total_reports: int
    grid_region: str

File: frontend/src/services/sustainability.ts
const API_BASE = '/api/v1/sustainability';

export interface APIResponse<T> {
  success: boolean;
  message: string;
  data: T;
}

export interface SustainabilitySummary {
  total_carbon_saved_kg: number;
  scope1_emission_kg: number;
  scope3_emission_kg: number;
  vehicles_assessed: number;
  average_readiness_score: number;
  total_reports: number;
  grid_region: string;
}

export interface CarbonReport {
  report_id: string;
  vehicle_id: string;
  distance_travelled: number;
  energy_consumed_kwh: number;
  diesel_emission: number;
  ev_emission: number;
  scope1_emission: number;
  scope3_emission: number;
  carbon_saved: number;
  grid_region: string;
  generated_at: string;
}

export interface DieselVsEVRequest {
  distance_km: number;
  payload_kg: number;
  diesel_efficiency: number;
  ev_efficiency: number;
  vehicle_type?: string;
}

export interface DieselVsEVResponse {
  diesel_emission: number;
  ev_emission: number;
  carbon_saved: number;
  reduction_percentage: number;
  equivalent_trees: number;
  equivalent_gallons_gasoline: number;
}

export interface ReadinessAssessmentRequest {
  route_distance: number;
  payload: number;
  charging_availability: boolean;
  dwell_time: number;
  vehicle_type?: string;
  vehicle_id?: string;
}

export interface ReadinessAssessmentResponse {
  assessment_id: string;
  readiness_score: number;
  readiness_level: string;
  recommendation: string;
  factor_scores: Record<string, number>;
  financial: Record<string, any>;
  improvements_needed: string[];
  generated_at: string;
}

export interface ProcurementRecommendationRequest {
  fleet_size: number;
  daily_distance: number;
  charging_available: boolean;
  vehicle_type?: string;
}

export interface ProcurementRecommendationResponse {
  recommended_vehicle_type: string;
  recommended_quantity: number;
  estimated_carbon_saving: number;
  recommendation_level: 'LOW' | 'MEDIUM' | 'HIGH';
  reasoning: string;
}

export const sustainabilityApi = {
  getSummary: async (): Promise<APIResponse<SustainabilitySummary>> => {
    const res = await fetch(`${API_BASE}/summary`);
    if (!res.ok) throw new Error('Failed to fetch sustainability summary');
    return res.json();
  },

  getHistory: async (): Promise<APIResponse<CarbonReport[]>> => {
    const res = await fetch(`${API_BASE}/history?limit=50`);
    if (!res.ok) throw new Error('Failed to fetch sustainability history');
    return res.json();
  },

  compareDieselVsEv: async (payload: DieselVsEVRequest): Promise<APIResponse<DieselVsEVResponse>> => {
    const res = await fetch(`${API_BASE}/diesel-vs-ev`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error('Failed to compare diesel vs ev');
    return res.json();
  },

  assessReadiness: async (payload: ReadinessAssessmentRequest): Promise<APIResponse<ReadinessAssessmentResponse>> => {
    const res = await fetch(`${API_BASE}/readiness-assessment`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error('Failed to assess readiness');
    return res.json();
  },

  getReadiness: async (assessmentId: string): Promise<APIResponse<ReadinessAssessmentResponse>> => {
    const res = await fetch(`${API_BASE}/readiness/${assessmentId}`);
    if (!res.ok) throw new Error('Failed to fetch readiness assessment');
    return res.json();
  },

  getProcurementRecommendation: async (payload: ProcurementRecommendationRequest): Promise<APIResponse<ProcurementRecommendationResponse>> => {
    const res = await fetch(`${API_BASE}/procurement-recommendation`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
    if (!res.ok) throw new Error('Failed to generate procurement recommendation');
    return res.json();
  },

  getCarbonReport: async (reportId: string): Promise<APIResponse<CarbonReport>> => {
    const res = await fetch(`${API_BASE}/carbon/report/${reportId}`);
    if (!res.ok) throw new Error('Failed to fetch carbon report');
    return res.json();
  }
};
